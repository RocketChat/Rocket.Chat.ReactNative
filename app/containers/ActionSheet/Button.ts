import { TouchableOpacity } from 'react-native';

import { isAndroid } from '../../utils/deviceInfo';
import Touch from '../../utils/touch';

// Taken from https://github.com/rgommezz/react-native-scroll-bottom-sheet#touchables
export const Button = isAndroid ? Touch : TouchableOpacity;
