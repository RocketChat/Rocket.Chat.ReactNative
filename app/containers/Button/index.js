import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text } from 'react-native';
import Touchable from 'react-native-platform-touchable';

import { themes } from '../../constants/colors';
import sharedStyles from '../../views/Styles';
import ActivityIndicator from '../ActivityIndicator';

const styles = StyleSheet.create({
	container: {
		paddingHorizontal: 15,
		justifyContent: 'center',
		height: 48,
		borderRadius: 2,
		marginBottom: 10
	},
	text: {
		fontSize: 18,
		textAlign: 'center'
	}
});

export default class Button extends React.PureComponent {
	static propTypes = {
		title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
		type: PropTypes.string,
		onPress: PropTypes.func,
		disabled: PropTypes.bool,
		backgroundColor: PropTypes.string,
		loading: PropTypes.bool,
		theme: PropTypes.string,
		style: PropTypes.any
	}

	static defaultProps = {
		title: 'Press me!',
		type: 'primary',
		onPress: () => alert('It works!'),
		disabled: false,
		loading: false
	}

	render() {
		const {
			title, type, onPress, disabled, backgroundColor, loading, style, theme, ...otherProps
		} = this.props;
		const isPrimary = type === 'primary';
		return (
			<Touchable
				onPress={onPress}
				disabled={disabled || loading}
				style={[
					styles.container,
					backgroundColor
						? { backgroundColor }
						: { backgroundColor: isPrimary ? themes[theme].actionTintColor : themes[theme].backgroundColor },
					disabled && { backgroundColor: themes[theme].borderColor },
					style
				]}
				{...otherProps}
			>
				{
					loading
						? <ActivityIndicator color={isPrimary ? themes[theme].buttonText : themes[theme].actionTintColor} />
						: (
							<Text
								style={[
									styles.text,
									isPrimary ? sharedStyles.textMedium : sharedStyles.textBold,
									{ color: isPrimary ? themes[theme].buttonText : themes[theme].actionTintColor }
								]}
							>
								{title}
							</Text>
						)
				}
			</Touchable>
		);
	}
}
