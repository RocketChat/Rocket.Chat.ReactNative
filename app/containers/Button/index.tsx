import React from 'react';
import { StyleProp, StyleSheet, Text, TextStyle } from 'react-native';
import Touchable, { PlatformTouchableProps } from 'react-native-platform-touchable';

import { useTheme } from '../../theme';
import sharedStyles from '../../views/Styles';
import ActivityIndicator from '../ActivityIndicator';

interface IButtonProps extends PlatformTouchableProps {
	title: string;
	onPress: () => void;
	type?: string;
	backgroundColor?: string;
	loading?: boolean;
	color?: string;
	fontSize?: number;
	styleText?: StyleProp<TextStyle> | StyleProp<TextStyle>[];
}

const styles = StyleSheet.create({
	container: {
		paddingHorizontal: 14,
		justifyContent: 'center',
		height: 48,
		borderRadius: 4,
		marginBottom: 12
	},
	text: {
		...sharedStyles.textMedium,
		...sharedStyles.textAlignCenter
	},
	disabled: {
		opacity: 0.3
	}
});

const Button = ({
	type = 'primary',
	disabled = false,
	loading = false,
	fontSize = 16,
	title,
	onPress,
	backgroundColor,
	color,
	style,
	styleText,
	...otherProps
}: IButtonProps): React.ReactElement => {
	const { colors } = useTheme();
	const isPrimary = type === 'primary';

	let textColor = isPrimary ? colors.buttonText : colors.bodyText;
	if (color) {
		textColor = color;
	}

	return (
		<Touchable
			onPress={onPress}
			disabled={disabled || loading}
			style={[
				styles.container,
				backgroundColor ? { backgroundColor } : { backgroundColor: isPrimary ? colors.actionTintColor : colors.backgroundColor },
				disabled && styles.disabled,
				style
			]}
			accessibilityLabel={title}
			{...otherProps}
		>
			{loading ? (
				<ActivityIndicator color={textColor} />
			) : (
				<Text style={[styles.text, { color: textColor, fontSize }, styleText]} accessibilityLabel={title}>
					{title}
				</Text>
			)}
		</Touchable>
	);
};

export default Button;
