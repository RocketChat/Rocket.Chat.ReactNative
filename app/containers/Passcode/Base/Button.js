import React from 'react';
import { Text } from 'react-native';
import PropTypes from 'prop-types';

import styles from './styles';
import { themes } from '../../../constants/colors';
import Touch from '../../../utils/touch';
import { CustomIcon } from '../../../lib/Icons';

const Button = React.memo(({
	text, disabled, theme, onPress, icon
}) => {
	const press = () => onPress && onPress(text);

	return (
		<Touch
			style={[styles.buttonView, { backgroundColor: 'transparent' }]}
			underlayColor={themes[theme].passcodeButtonActive}
			rippleColor={themes[theme].passcodeButtonActive}
			enabled={!disabled}
			theme={theme}
			onPress={press}
		>
			{
				icon
					? (
						<CustomIcon name={icon} size={36} color={themes[theme].passcodePrimary} />
					)
					: (
						<Text style={[styles.buttonText, { color: themes[theme].passcodePrimary }]}>
							{text}
						</Text>
					)
			}
		</Touch>
	);
});

Button.propTypes = {
	text: PropTypes.string,
	icon: PropTypes.string,
	theme: PropTypes.string,
	disabled: PropTypes.bool,
	onPress: PropTypes.func
};

export default Button;
