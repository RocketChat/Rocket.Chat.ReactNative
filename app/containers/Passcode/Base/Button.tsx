import React from 'react';
import { Text, StyleProp, ViewStyle } from 'react-native';

import styles from './styles';
import { themes } from '../../../lib/constants';
import Touch from '../../Touch';
import { CustomIcon, TIconsName } from '../../CustomIcon';
import { useTheme } from '../../../theme';

interface IPasscodeButton {
	text?: string;
	icon?: TIconsName;
	disabled?: boolean;
	onPress?: Function;
	style?: StyleProp<ViewStyle>;
}

const Button = React.memo(({ style, text, disabled, onPress, icon }: IPasscodeButton) => {
	const { theme } = useTheme();

	const press = () => onPress && onPress(text);

	return (
		<Touch
			style={[styles.buttonView, { backgroundColor: 'transparent' }, style]}
			underlayColor={themes[theme].passcodeButtonActive}
			rippleColor={themes[theme].passcodeButtonActive}
			enabled={!disabled}
			onPress={press}
		>
			{icon ? (
				<CustomIcon name={icon} size={36} color={themes[theme].passcodePrimary} />
			) : (
				<Text style={[styles.buttonText, { color: themes[theme].passcodePrimary }]}>{text}</Text>
			)}
		</Touch>
	);
});

export default Button;
