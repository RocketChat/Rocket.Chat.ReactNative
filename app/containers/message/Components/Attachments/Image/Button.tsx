import React from 'react';

import { useTheme } from '../../../../../theme';
import Touchable from '../../../Touchable';
import styles from '../../../styles';

interface IMessageButton {
	children: React.ReactElement;
	disabled?: boolean;
	onPress: () => void;
}

export const Button = ({ children, onPress, disabled }: IMessageButton) => {
	const { colors } = useTheme();
	return (
		<Touchable
			disabled={disabled}
			onPress={onPress}
			style={styles.imageContainer}
			background={Touchable.Ripple(colors.surfaceNeutral)}>
			{children}
		</Touchable>
	);
};
