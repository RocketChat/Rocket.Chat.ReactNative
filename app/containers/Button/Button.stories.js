import React from 'react';
import { storiesOf } from '@storybook/react-native';

import Button from '.';

const buttonProps = {
	title: 'Press me!',
	type: 'primary',
	onPress: () => {},
	testID: 'testButton',
	fontSize: 16,
	style: {
		padding: 10,
		justifyContent: 'center'
	}
};

const stories = storiesOf('Button', module);

stories.add('primary button', () => <Button {...buttonProps} />);

stories.add('secondary button', () => <Button {...buttonProps} type='secondary' />);

stories.add('loading button', () => <Button loading {...buttonProps} />);

stories.add('disabled button', () => <Button disabled {...buttonProps} />);

stories.add('disabled loading button', () => <Button disabled loading {...buttonProps} />);
