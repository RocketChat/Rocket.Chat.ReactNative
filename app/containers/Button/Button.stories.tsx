import React from 'react';

import Button from '.';

const buttonProps = {
	title: 'Press me!',
	type: 'primary',
	onPress: () => {},
	testID: 'testButton'
};

export default {
	title: 'Button'
};

export const PrimaryButton = () => <Button {...buttonProps} />;

export const SecondaryButton = () => <Button {...buttonProps} type='secondary' />;

export const LoadingButton = () => <Button loading {...buttonProps} />;

export const DisabledButton = () => <Button disabled {...buttonProps} />;

export const DisabledLoadingButton = () => <Button disabled loading {...buttonProps} />;

export const CustomButton = () => (
	<Button
		{...buttonProps}
		fontSize={18}
		backgroundColor='purple'
		color='yellow'
		style={{
			padding: 10
		}}
		styleText={[
			{
				textAlign: 'left'
			}
		]}
	/>
);
