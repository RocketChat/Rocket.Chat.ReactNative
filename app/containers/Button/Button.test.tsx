import React from 'react';
import { View } from 'react-native';
import { fireEvent, render } from '@testing-library/react-native';

import Button from '.';

const onPressMock = jest.fn();

const testProps = {
	title: 'Press me!',
	type: 'primary',
	onPress: onPressMock,
	testID: 'testButton',
	initialText: 'Initial text',
	textAfterPress: 'Button pressed!'
};

const TestButton = ({ loading = false, disabled = false }) => (
	<View>
		<Button
			title={testProps.title}
			type={testProps.title}
			onPress={testProps.onPress}
			testID={testProps.testID}
			accessibilityLabel={testProps.title}
			disabled={disabled}
			loading={loading}
		/>
	</View>
);

describe('ButtonTests', () => {
	test('rendered', async () => {
		const { findByTestId } = render(<TestButton />);
		const Button = await findByTestId(testProps.testID);
		expect(Button).toBeTruthy();
	});

	test('rendered with correct title', async () => {
		const { findByText } = render(<TestButton />);
		const ButtonTitle = await findByText(testProps.title);
		expect(ButtonTitle).toBeTruthy();
		expect(ButtonTitle.props.children).toEqual(testProps.title);
	});

	test('find button using accessibilityLabel', async () => {
		const { findByA11yLabel } = render(<TestButton />);
		const Button = await findByA11yLabel(testProps.title);
		expect(Button).toBeTruthy();
	});

	test('title not visible while loading', async () => {
		const { queryByText } = render(<TestButton loading={true} />);
		const ButtonTitle = await queryByText(testProps.title);
		expect(ButtonTitle).toBeNull();
	});

	test('should not trigger onPress on disabled button', async () => {
		const { findByTestId } = render(<TestButton disabled={true} />);
		const Button = await findByTestId(testProps.testID);
		fireEvent.press(Button);
		expect(onPressMock).not.toHaveBeenCalled();
	});

	test('should trigger onPress function on button press', async () => {
		const { findByTestId } = render(<TestButton />);
		const Button = await findByTestId(testProps.testID);
		fireEvent.press(Button);
		expect(onPressMock).toHaveBeenCalled();
	});
});
