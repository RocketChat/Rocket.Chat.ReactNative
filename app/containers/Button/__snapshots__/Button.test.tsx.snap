// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story Snapshots: CustomButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "purple",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "padding": 10,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <Text
    style={
      [
        {
          "color": "yellow",
          "fontSize": 18,
        },
        {
          "backgroundColor": "transparent",
          "fontFamily": "Inter",
          "fontWeight": "500",
          "textAlign": "center",
        },
        [
          {
            "textAlign": "left",
          },
        ],
      ]
    }
  >
    Press me!
  </Text>
</View>
`;

exports[`Story Snapshots: DisabledButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#D1EBFE",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
        },
        {
          "backgroundColor": "transparent",
          "fontFamily": "Inter",
          "fontWeight": "500",
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;

exports[`Story Snapshots: DisabledLoadingButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#D1EBFE",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <ActivityIndicator
    color="#FFFFFF"
    style={
      {
        "padding": 0,
      }
    }
  />
</View>
`;

exports[`Story Snapshots: LoadingButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#D1EBFE",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <ActivityIndicator
    color="#FFFFFF"
    style={
      {
        "padding": 0,
      }
    }
  />
</View>
`;

exports[`Story Snapshots: PrimaryButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#156FF5",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
        },
        {
          "backgroundColor": "transparent",
          "fontFamily": "Inter",
          "fontWeight": "500",
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;

exports[`Story Snapshots: SecondaryButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#E4E7EA",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 14,
    }
  }
  testID="testButton"
>
  <Text
    style={
      [
        {
          "color": "#2F343D",
          "fontSize": 16,
        },
        {
          "backgroundColor": "transparent",
          "fontFamily": "Inter",
          "fontWeight": "500",
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;

exports[`Story Snapshots: SmallButton should match snapshot 1`] = `
<View
  accessibilityLabel="Press me!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "center",
      "backgroundColor": "#156FF5",
      "borderRadius": 4,
      "marginBottom": 12,
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
  testID="testButton"
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
        },
        {
          "backgroundColor": "transparent",
          "fontFamily": "Inter",
          "fontSize": 12,
          "fontWeight": "700",
          "lineHeight": 18,
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;
